### PR TITLE
getdisplaymedia: add displaysurface

### PR DESCRIPTION
**Pull request by fippo**
described in https://developer.chrome.com/docs/web-platform/screen-sharing-controls/

@eladalon1983 it seems https://developer.chrome.com/docs/web-platform/screen-sharing-controls/#displaySurface has displaySurface at the wrong level, no?

### CODE REVIEW

**Comment by eladalon1983** [comment]
> @eladalon1983 it seems https://developer.chrome.com/docs/web-platform/screen-sharing-controls/#displaySurface has displaySurface at the wrong level, no?

Yes, that's wrong atm. Thanks for pointing it out! The article will be fixed soon.

**Comment by beaufortfrancois** [comment]
Thanks @fippo for raising this issue! We're fixing the article in https://github.com/GoogleChrome/developer.chrome.com/pull/3881

File: src/content/getusermedia/getdisplaymedia/index.html [hunk 1]
```diff
@@ -35,7 +35,12 @@ <h1><a href="//webrtc.github.io/samples/" title="WebRTC samples homepage">WebRTC
 
     <video id="gum-local" autoplay playsinline muted></video>
     <button id="startButton" disabled>Start</button>
-
+    <select id="preference" style="display:none">
+      <option value="" selected>Share the default</option>
+      <option value="browser">Share a browser window</option>
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/index.html]:
```suggestion
      <option value="browser">Share a browser tab</option>
```

File: src/content/getusermedia/getdisplaymedia/index.html [hunk 2]
```diff
@@ -35,7 +35,12 @@ <h1><a href="//webrtc.github.io/samples/" title="WebRTC samples homepage">WebRTC
...
+      <option value="window">Share another window</option>
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/index.html]:
```suggestion
      <option value="window">Share a window</option>
```

File: src/content/getusermedia/getdisplaymedia/js/main.js [hunk 3]
```diff
@@ -7,9 +7,14 @@
  */
 'use strict';
 
-// Polyfill in Firefox.
-// See https://blog.mozilla.org/webrtc/getdisplaymedia-now-available-in-adapter-js/
-if (adapter.browserDetails.browser == 'firefox') {
+const sharingPreference = document.getElementById('preference');
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/js/main.js]: How about `preferredDisplaySurface` instead of `sharingPreference` so that it's clear?

File: src/content/getusermedia/getdisplaymedia/js/main.js [hunk 4]
```diff
@@ -7,9 +7,14 @@
...
+
+if (adapter.browserDetails.browser === 'chrome' && adapter.browserDetails.version >= 107) {
+  // See https://developer.chrome.com/docs/web-platform/screen-sharing-controls/
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/js/main.js:15]: I'd prefer not to promote this pattern. How about simply letting the `select` always visible and add a link to the article or MDN later?

**Review by beaufortfrancois — COMMENTED** [review]
I've left few comments!

Note that https://developer.chrome.com/docs/web-platform/screen-sharing-controls/#displaySurface has been updated ;)

File: src/content/getusermedia/getdisplaymedia/index.html (hunk 2 above)
Comment (eladalon1983) [reply src/content/getusermedia/getdisplaymedia/index.html]: Why "browser window"? It can be any native window.

Comment (eladalon1983) [reply src/content/getusermedia/getdisplaymedia/index.html]: ("Another" is also wrong, btw; one can share the current window too.)

File: src/content/getusermedia/getdisplaymedia/js/main.js (hunk 4 above)
Comment (fippo) [reply src/content/getusermedia/getdisplaymedia/js/main.js:15]: this pattern is the sad reality (also here in the samples) unless there is a way to feature detect :-(

Comment (beaufortfrancois) [reply src/content/getusermedia/getdisplaymedia/js/main.js:15]: 👍

File: src/content/getusermedia/getdisplaymedia/index.html (hunk 2 above)
Comment (beaufortfrancois) [reply src/content/getusermedia/getdisplaymedia/index.html]: Good catch. I've edited my comment to "Share a window"

File: src/content/getusermedia/getdisplaymedia/js/main.js (hunk 3 above)
Comment (fippo) [reply src/content/getusermedia/getdisplaymedia/js/main.js]: and a better name for the element

File: src/content/getusermedia/getdisplaymedia/js/main.js [hunk 5]
```diff
@@ -7,14 +7,20 @@
  */
 'use strict';
 
-// Polyfill in Firefox.
-// See https://blog.mozilla.org/webrtc/getdisplaymedia-now-available-in-adapter-js/
-if (adapter.browserDetails.browser == 'firefox') {
+const preferredDisplaySurface = document.getElementById('displaySurface');
+
+if (adapter.browserDetails.browser === 'chrome' && adapter.browserDetails.version >= 107) {
```
Comment (eladalon1983) [inline src/content/getusermedia/getdisplaymedia/js/main.js]: nit: Mind breaking up this line?

File: src/content/getusermedia/getdisplaymedia/js/main.js [hunk 6]
```diff
@@ -40,7 +47,12 @@ function errorMsg(msg, error) {
 
 const startButton = document.getElementById('startButton');
 startButton.addEventListener('click', () => {
-  navigator.mediaDevices.getDisplayMedia({video: true, audio: true})
+  const options = {audio: true, video: {}};
+  const displaySurface = preferredDisplaySurface.options[preferredDisplaySurface.selectedIndex];
+  if (displaySurface.value !== '') {
```
Comment (eladalon1983) [inline src/content/getusermedia/getdisplaymedia/js/main.js]: nit: Is this the idiomatic thing to do? What about `if (displaySurface.value) {`?

**Review by eladalon1983 — APPROVED** [review]

**Comment by eladalon1983** [comment]
Generally LGTM, but I wonder if adding more than the minimal functionality to an example, might make it less approachable. Please consider if this might deserve its own sample page, possibly along with other similar controls, like in François' page [here](https://screen-sharing-controls.glitch.me/).

File: src/content/getusermedia/getdisplaymedia/js/main.js (hunk 6 above)
Comment (fippo) [reply src/content/getusermedia/getdisplaymedia/js/main.js]: idiomatic yes, good for a sample no

File: src/content/getusermedia/getdisplaymedia/js/main.js (hunk 5 above)
Comment (beaufortfrancois) [reply src/content/getusermedia/getdisplaymedia/js/main.js]: +1
```suggestion
if (adapter.browserDetails.browser === 'chrome' && 
    adapter.browserDetails.version >= 107) {
```

**Review by beaufortfrancois — APPROVED** [review]
LGTM with nit

https://webrtc.github.io/samples/src/content/getusermedia/getdisplaymedia/ is a canonical page for getDisplayMedia. I think it's fine to have `displaySurface` there.

**Comment by eladalon1983** [comment]
> https://webrtc.github.io/samples/src/content/getusermedia/getdisplaymedia/ is a canonical page for getDisplayMedia. I think it's fine to have `displaySurface` there.

I think it's fine but sub-optimal. I think ideally, this page should have the minimal required functionality.
* Beginners can make sense of it more easily.
* When people are sent there for debugging purposes, they don't end up puzzled over irrelevant controls. (I often use "does the problem manifest on \*this\* page too?")

**Comment by beaufortfrancois** [comment]
How about bundling those advanced options like this?

image


```diff
 
     <video id="gum-local" autoplay playsinline muted></video>
+    <fieldset>
+      <legend>Advanced options</legend>
+      <select id="preference" style="display:none">
+        <option value="" selected>Share the default</option>
+        <option value="browser">Share a browser window</option>
+        <option value="window">Share another window</option>
+        <option value="monitor">Share the entire screen</option>
+      </select>
+    </fieldset>
     <button id="startButton" disabled>Start</button>
-    <select id="preference" style="display:none">
-      <option value="" selected>Share the default</option>
-      <option value="browser">Share a browser window</option>
-      <option value="window">Share another window</option>
-      <option value="monitor">Share the entire screen</option>
-    </select>
```

**Comment by eladalon1983** [comment]
> How about bundling those advanced options like this?
> ...

Yeah, that makes sense to me. Especially if the example can still demonstrate that `video: true` is an option. (I think a lot of developers end up not needing constraints. But I've not checked this hypothesis.)

**Comment by beaufortfrancois** [comment]
@fippo How does that sound? If that's too much, I can take this PR over. Let me know.

File: src/content/getusermedia/getdisplaymedia/js/main.js (hunk 6 above)
Comment (fippo) [reply src/content/getusermedia/getdisplaymedia/js/main.js]: changed to `!= 'default'` which is easier to understand

**Comment by fippo** [comment]
I like the fieldset! Borrowed with updated labels. Also changed the labels to "prefer to share"

File: src/content/getusermedia/getdisplaymedia/index.html [hunk 7]
```diff
@@ -35,7 +35,15 @@ <h1><a href="//webrtc.github.io/samples/" title="WebRTC samples homepage">WebRTC
 
     <video id="gum-local" autoplay playsinline muted></video>
     <button id="startButton" disabled>Start</button>
-
+    <fieldset id="options" style="display:none">
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/index.html:38]: You may want to update CSS with something like:

```diff
+fieldset {
+  margin: 0 0 1em 0;
+}
+
+fieldset > select {
+  margin-top: 1em;
+}
+
```

File: src/content/getusermedia/getdisplaymedia/index.html [hunk 8]
```diff
@@ -35,7 +35,15 @@ <h1><a href="//webrtc.github.io/samples/" title="WebRTC samples homepage">WebRTC
...
+      <legend>Advanced options</legend>
+      <select id="displaySurface">
+        <option value="default" selected>Show default sharing options</option>
+        <option value="browser">Prefer to share a browser tab</option>
+        <option value="window">Prefer to share a window</option>
+        <option value="monitor">Prefer to share the entire screen</option>
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/index.html]:
```suggestion
        <option value="monitor">Prefer to share an entire screen</option>
```
There can be multiple screens plugged.

File: src/content/getusermedia/getdisplaymedia/js/main.js [hunk 9]
```diff
@@ -40,7 +48,12 @@ function errorMsg(msg, error) {
 
 const startButton = document.getElementById('startButton');
 startButton.addEventListener('click', () => {
-  navigator.mediaDevices.getDisplayMedia({video: true, audio: true})
+  const options = {audio: true, video: {}};
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/js/main.js]: As we may be updating `options` after, we should use `let`.
```suggestion
  let options = {audio: true, video: true};
```

File: src/content/getusermedia/getdisplaymedia/js/main.js [hunk 10]
```diff
@@ -40,7 +48,12 @@ function errorMsg(msg, error) {
...
+  const displaySurface = preferredDisplaySurface.options[preferredDisplaySurface.selectedIndex];
+  if (displaySurface.value !== 'default') {
+    options.video.displaySurface = displaySurface.value;
+  }
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/js/main.js]:
```suggestion
  const displaySurface = preferredDisplaySurface.options[preferredDisplaySurface.selectedIndex].value;
  if (displaySurface !== 'default') {
    options.video = { displaySurface };
  }
```

File: src/content/getusermedia/getdisplaymedia/index.html (hunk 7 above)
Comment (fippo) [reply src/content/getusermedia/getdisplaymedia/index.html:38]: added to main css, tagging this as TBR=@samdutton ;-)

File: src/content/getusermedia/getdisplaymedia/js/main.js (hunk 9 above)
Comment (fippo) [reply src/content/getusermedia/getdisplaymedia/js/main.js]: JS semantics: you can change properties of a const map 🤷‍♂️

Comment (eladalon1983) [reply src/content/getusermedia/getdisplaymedia/js/main.js]: > JS semantics: you can change properties of a const map 🤷‍♂️

I usually think of it as the difference between a const-ptr and a ptr-to-const, then it makes sense.

File: src/content/getusermedia/getdisplaymedia/js/main.js [hunk 11]
```diff
@@ -38,9 +47,14 @@ function errorMsg(msg, error) {
   }
 }
 
-const startButton = document.getElementById('startButton');
```
Comment (beaufortfrancois) [inline src/content/getusermedia/getdisplaymedia/js/main.js:41]: Why did you remove this line?

**Review by beaufortfrancois — APPROVED** [review]

**Comment by beaufortfrancois** [comment]
LGTM with nit

@fippo I'll merge after `const startButton = document.getElementById('startButton');` line is restored if that looks good to you.

image

File: src/content/getusermedia/getdisplaymedia/js/main.js (hunk 11 above)
Comment (fippo) [reply src/content/getusermedia/getdisplaymedia/js/main.js:41]: not removed, moved up. We do declare the the UI-related accessors at the top

**Comment by beaufortfrancois** [comment]
And it's live at https://webrtc.github.io/samples/src/content/getusermedia/getdisplaymedia/! Thank you @fippo!

image
